File: bootstrap/src/main/java/example/greenwayb/bootstrap/BootstrapApplicationListener.java
package example.greenwayb.bootstrap;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.Ordered;
import org.springframework.core.env.CompositePropertySource;
import org.springframework.core.env.ConfigurableEnvironment;
import org.springframework.core.io.support.SpringFactoriesLoader;

import java.util.List;

public class BootstrapApplicationListener implements ApplicationContextInitializer, Ordered {

    public static final String PROPERTY_NAME = "greenwaybProperties";
    private static boolean initialized;
    private static final Log LOG = LogFactory.getLog(BootstrapApplicationListener.class);

    @Override
    public void initialize(ConfigurableApplicationContext applicationContext) {
        if (!initialized) {
            ConfigurableEnvironment environment = applicationContext.getEnvironment();
            List<PropertySourceLocator> factories = SpringFactoriesLoader.loadFactories(PropertySourceLocator.class, null);

            CompositePropertySource ps = (CompositePropertySource)environment.getPropertySources().get(PROPERTY_NAME);
            if (ps == null) {
                ps = new CompositePropertySource(PROPERTY_NAME);
                environment.getPropertySources().addLast(ps);
            }

            for (PropertySourceLocator factory : factories){
                ps.addPropertySource(factory.locate(environment));
            }

            initialized = true;
            LOG.info("Property sources are " + environment.getPropertySources());
        }
    }

    @Override
    public int getOrder() {
        return Ordered.HIGHEST_PRECEDENCE;
    }
}

File: bootstrap/src/main/kotlin/example/greenwayb/bootstrap/BootstrapApplicationListener.kt
package example.greenwayb.bootstrap

import com.github.greenwayb.bootstrap.PropertySourceLocator
import org.apache.commons.logging.Log
import org.apache.commons.logging.LogFactory
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.core.Ordered
import org.springframework.core.env.CompositePropertySource
import org.springframework.core.io.support.SpringFactoriesLoader

open class BootstrapApplicationListener : ApplicationContextInitializer<ConfigurableApplicationContext>, Ordered {

    companion object {
        private val LOG : Log = LogFactory.getLog(BootstrapApplicationListener::class.java)

        const val PROPERTY_NAME : String = "greenwaybProperties"

        private var initalized = false
    }

    override fun getOrder(): Int {
        return Ordered.HIGHEST_PRECEDENCE
    }

    override fun initialize(applicationContext: ConfigurableApplicationContext) {
        if (!initalized) {
            val environment = applicationContext.environment

            val factories = SpringFactoriesLoader
                    .loadFactories(PropertySourceLocator::class.java, null)

            var ps = environment.propertySources.get(PROPERTY_NAME) as CompositePropertySource?
            if (null == ps) {
                ps = CompositePropertySource(PROPERTY_NAME)
                environment.propertySources.addLast(ps)
            }

            for (factory in factories) {
                ps.addPropertySource(factory.locate(environment))
            }
            initalized = true
            LOG.info("Property sources are ${environment.propertySources}")
        }
    }
}
